pkg/controllers/updaterun/stop.go

```diff
@@ -0,0 +1,237 @@
+	klog.InfoS("The updating stage is waiting for updating clusters to finish before completely stopping", "numberOfUpdatingClusters", clusterUpdatingCount, "stage", updatingStageStatus.StageName, "updateRun", updateRunRef)
+	return false, clusterUpdatingWaitTime, nil
+}
+
+// stopDeleteStage stops the delete stage by letting the deleting bindings finish.
+func (r *Reconciler) stopDeleteStage(
+	updateRun placementv1beta1.UpdateRunObj,
+	toBeDeletedBindings []placementv1beta1.BindingObj,
+) (bool, error) {
+	updateRunRef := klog.KObj(updateRun)
+	updateRunStatus := updateRun.GetUpdateRunStatus()
+	existingDeleteStageStatus := updateRunStatus.DeletionStageStatus
+	existingDeleteStageClusterMap := make(map[string]*placementv1beta1.ClusterUpdatingStatus, len(existingDeleteStageStatus.Clusters))
+	for i := range existingDeleteStageStatus.Clusters {
+		existingDeleteStageClusterMap[existingDeleteStageStatus.Clusters[i].ClusterName] = &existingDeleteStageStatus.Clusters[i]
+	}
+	// Mark the delete stage as stopping in case it's not.
+	markStageUpdatingStopping(existingDeleteStageStatus, updateRun.GetGeneration())
+
+	for _, binding := range toBeDeletedBindings {
+		bindingSpec := binding.GetBindingSpec()
+		curCluster, exist := existingDeleteStageClusterMap[bindingSpec.TargetCluster]
+		if !exist {
+			// This is unexpected because we already checked in validation.
+			missingErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the to be deleted cluster `%s` is not in the deleting stage during stopping", bindingSpec.TargetCluster))
+			klog.ErrorS(missingErr, "The cluster in the deleting stage does not include all the to be deleted binding", "updateRun", updateRunRef)
+			return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, missingErr.Error())
+		}
+		// In validation, we already check the binding must exist in the status.
+		delete(existingDeleteStageClusterMap, bindingSpec.TargetCluster)
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(curCluster.Conditions, string(placementv1beta1.ClusterUpdatingConditionSucceeded)), updateRun.GetGeneration()) {
+			// The cluster status is marked as deleted.
+			continue
+		}
+		if condition.IsConditionStatusTrue(meta.FindStatusCondition(curCluster.Conditions, string(placementv1beta1.ClusterUpdatingConditionStarted)), updateRun.GetGeneration()) {
+			// The cluster status is marked as being deleted.
+			if binding.GetDeletionTimestamp().IsZero() {
+				// The cluster is marked as deleting but the binding is not deleting.
+				unexpectedErr := controller.NewUnexpectedBehaviorError(fmt.Errorf("the cluster `%s` in the deleting stage is marked as deleting but its corresponding binding is not deleting", curCluster.ClusterName))
+				klog.ErrorS(unexpectedErr, "The binding should be deleting before we mark a cluster deleting", "clusterStatus", curCluster, "updateRun", updateRunRef)
+				return false, fmt.Errorf("%w: %s", errStagedUpdatedAborted, unexpectedErr.Error())
+			}
+			return false, nil
```

Comment:
Why do we return here ?, we just know that this binding is deleting at the moment why not check other toBeDeletedBindings ?